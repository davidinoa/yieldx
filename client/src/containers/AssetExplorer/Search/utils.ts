interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
