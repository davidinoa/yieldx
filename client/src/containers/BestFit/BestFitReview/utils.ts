import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
