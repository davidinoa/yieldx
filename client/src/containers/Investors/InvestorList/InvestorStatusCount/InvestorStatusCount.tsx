import React from "react";
import { AccountStatus } from "@bondhouse/apex";
import { Row } from "react-bootstrap";
import { useRouteMatch } from "react-router-dom";
import {
  CreateInvestorButton,
  ListFilters,
  ListStatuses,
  Status
} from "../../../../components/ListPages/ListPages.styles";
import { Spinner } from "../../../../elements/Spinner/Spinner";

interface Props {
  canCreateAccount: boolean;
  statusCounts?: {
    pending: number;
    active: number;
    inactive: number;
    error: number;
  };
}

export default function InvestorStatusCount({
  canCreateAccount,
  statusCounts
}: Props) {
  const { url } = useRouteMatch();

  return (
    <Row>
      <ListStatuses>
        <span>
          <Status className="pending" />
          <b>Pending:</b>
          {statusCounts === undefined ? (
            <span>
              <Spinner />
            </span>
          ) : (
            statusCounts.pending
          )}
        </span>
        <span>
          <Status className="active" />
          <b>Active:</b>
          {statusCounts === undefined ? (
            <span>
              <Spinner />
            </span>
          ) : (
            statusCounts.active
          )}
        </span>
        <span>
          <Status className="inactive" />
          <b>Inactive:</b>
          {statusCounts === undefined ? (
            <span>
              <Spinner />
            </span>
          ) : (
            statusCounts.inactive
          )}
        </span>
        <span>
          <Status className={`${AccountStatus.ERROR.toLowerCase()}`} />
          <b>ISSUES:</b>
          {statusCounts === undefined ? (
            <span>
              <Spinner />
            </span>
          ) : (
            statusCounts.error
          )}
        </span>
      </ListStatuses>
      {canCreateAccount && (
        <ListFilters>
          <CreateInvestorButton
            data-testid="InvestorList-CreateInvestorLink"
            to={`${url}/create/investor-type`}
          >
            Create investor account
          </CreateInvestorButton>
        </ListFilters>
      )}
    </Row>
  );
}
