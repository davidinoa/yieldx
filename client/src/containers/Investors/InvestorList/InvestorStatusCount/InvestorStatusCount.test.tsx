import React from "react";
import InvestorStatusCount from "./InvestorStatusCount";
import { render, screen } from "../../../../test/utils";

describe("InvestorStatusCount", () => {
  test("Should render", async () => {
    const { container } = render(<InvestorStatusCount canCreateAccount />);
    expect(container).toBeInTheDocument();
  });

  test("Should render correct status labels", async () => {
    render(<InvestorStatusCount canCreateAccount />);
    expect(screen.getByText(/^PENDING:$/i)).toBeInTheDocument();
    expect(screen.getByText(/^ACTIVE:$/i)).toBeInTheDocument();
    expect(screen.getByText(/^INACTIVE:$/i)).toBeInTheDocument();
    expect(screen.getByText(/^ISSUES:$/i)).toBeInTheDocument();
  });

  test("Should render spinners if totals undefined", async () => {
    render(<InvestorStatusCount canCreateAccount />);
    expect(screen.getAllByTestId("spinner").length).toBe(4);
  });

  test("Should render create investor button according to canCreateAccount prop", async () => {
    const { rerender } = render(<InvestorStatusCount canCreateAccount />);
    expect(
      screen.queryByTestId("InvestorList-CreateInvestorLink")
    ).toBeInTheDocument();
    rerender(<InvestorStatusCount canCreateAccount={false} />);
    expect(
      screen.queryByTestId("InvestorList-CreateInvestorLink")
    ).not.toBeInTheDocument();
  });

  test("Should render status counts properly", async () => {
    const mockStatusCounts = {
      active: 6,
      pending: 0,
      inactive: 0,
      error: 2
    };
    const { asFragment } = render(
      <InvestorStatusCount canCreateAccount statusCounts={mockStatusCounts} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
