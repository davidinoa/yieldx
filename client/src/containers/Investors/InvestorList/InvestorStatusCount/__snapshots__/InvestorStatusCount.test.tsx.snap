// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestorStatusCount Should render status counts properly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-top: 0.35rem;
  padding-bottom: 0.35rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 span {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  position: relative;
  padding-right: 0.75rem;
  margin-right: 0.75rem;
  font-size: 0.65rem;
}

.c0 span i {
  display: inline-block;
  margin-right: 0.25rem;
}

.c0 span b {
  margin-right: 5px;
  margin-left: 5px;
  font-weight: bold;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--subtext-3);
}

.c0 span::after {
  content: "";
  border-right: solid 1px var(--light);
  width: 2px;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
}

.c0 span:last-of-type::after {
  border: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c2 .custom-control.custom-checkbox .custom-control-.label {
  padding-top: 5px;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
}

.c2 .custom-control.custom-checkbox .custom-control-.label::before {
  background: transparent;
  border: 1.5px solid var(--subtext-3);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5px;
  width: 5px;
  border-radius: 50%;
  margin: 0 auto;
}

.c1.active,
.c1.ready {
  background: var(--active);
}

.c1.deactivated,
.c1.inactive,
.c1.blocked,
.c1.closed,
.c1.terminated {
  background: var(--subtext-3);
}

.c1.tracking {
  background: var(--tracking);
}

.c1.pending {
  background: var(--caution);
}

.c1.error {
  background: var(--issue);
}

.c3 {
  text-transform: capitalize;
  font-weight: bold;
  border-radius: 0.25rem;
  padding: 0.55rem 2.15rem;
  font-size: 0.9rem;
  line-height: normal;
  white-space: nowrap;
  border: 1px solid var(--border-new);
  background: none;
  color: var(--primary);
}

.c3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  border: 1px solid var(--border-3);
}

.c3.empty-list {
  background: var(--primary);
  color: var(--white);
  border-color: var(--primary);
}

@media (min-width:1600px) {
  .c0 {
    padding-top: 1rem;
    padding-bottom: 1rem;
  }
}

@media (min-width:1600px) {
  .c0 span {
    padding-right: 1.25rem;
    margin-right: 1.25rem;
    font-size: 0.75rem;
  }
}

@media (min-width:1600px) {
  .c1 {
    height: 6px;
    width: 6px;
  }
}

@media (min-width:1600px) {
  .c3 {
    padding: 0.65rem 2.75rem;
    font-size: 0.95rem;
  }
}

<div
    class="row"
  >
    <div
      class="c0 col-17 col-xxl-16"
    >
      <span>
        <i
          class="c1 pending"
        />
        <b>
          Pending:
        </b>
        0
      </span>
      <span>
        <i
          class="c1 active"
        />
        <b>
          Active:
        </b>
        6
      </span>
      <span>
        <i
          class="c1 inactive"
        />
        <b>
          Inactive:
        </b>
        0
      </span>
      <span>
        <i
          class="c1 error"
        />
        <b>
          ISSUES:
        </b>
        2
      </span>
    </div>
    <div
      class="c2 col-7 col-xxl-8"
    >
      <a
        class="c3"
        data-testid="InvestorList-CreateInvestorLink"
        href="//create/investor-type"
      >
        Create investor account
      </a>
    </div>
  </div>
</DocumentFragment>
`;
